ipa-core/src/ff/boolean_array.rs

```diff
@@ -41,9 +48,28 @@ impl<A> BooleanArray for A where
         + ArrayAccess<Output = Boolean>
         + Expand<Input = Boolean>
         + FromIterator<Boolean>
+        + TryInto<Vec<Gf32Bit>, Error = crate::error::Error>
 {
 }
 
+impl<A> TryInto<Vec<AdditiveShare<Gf32Bit>>> for &AdditiveShare<A>
+where
+    A: BooleanArray,
+    AdditiveShare<A>: Sized,
+{
+    type Error = Error;
+
+    fn try_into(self) -> Result<Vec<AdditiveShare<Gf32Bit>>, Self::Error> {
+        let left_shares: Vec<Gf32Bit> = self.left().try_into()?;
```

Comment:
it is sad that we need 3 allocations here - I missed that. 

Is there a chance we could define a function on `BA` type that returns an iterator over `Gf32Bit` values instead of defining `TryFrom` implementation? This won't require trait bounds and callers can decide when to allocate. In this case, we would just zip two iterators and allocate at the end

Reply:
I see. That makes sense and should work.